### PR TITLE
fixup algs contd 3

### LINKED ISSUE #466: [[Create]](options) method does not need to call authenticatorMakeCredential() with separate |rpId|
`[[Create]](options)` method does not need to call authenticatorMakeCredential() with separate |rpId| because it is called with `options.rp` of type PublicKeyCredentialEntity whose `id` member _is_ the RP ID (as is noted in the changes queued as a part of PR #464).

### LINKED ISSUE #472: credential ID returned by authenticatorGetAssertion() is optional if allowList has exactly one member
[The authenticatorGetAssertion operation](https://www.w3.org/TR/webauthn/#op-get-assertion) says:
> On successful completion, the authenticator returns to the user agent:
    * **_The identifier of the credential used to generate the signature_**.
    * The authenticator data used to generate the signature.
    * The assertion signature.

However, the [CTAP spec](https://fidoalliance.org/specs/fido-v2.0-rd-20161004/fido-client-to-authenticator-protocol-v2.0-rd-20161004.html) says in [4.2 authenticatorGetAssertion](https://fidoalliance.org/specs/fido-v2.0-rd-20161004/fido-client-to-authenticator-protocol-v2.0-rd-20161004.html#authenticatorgetassertion): 
>On success, the authenticator must return the following structure in its response:
```
Member name  Data type   Required?  Definition
credential   Credential  Optional   Credential whose private key was used to 
                                     generate the assertion. May be
                                     omitted if the whitelist has exactly 
                                     one Credential.
[...]
```
..which would seem to be an optimization for CTAP where it does not have to return as many bytes (in what may be a common case).

Update WebAuthn to reflect this?

### LINKED ISSUE #536: parameter lists in #createCredential and #op-make-cred do not match
[#createCredential](https://www.w3.org/TR/webauthn/#createCredential) is missing `options.authenticatorSelection.uv` in step 24.3, and is in different order than [#op-make-cred](https://www.w3.org/TR/webauthn/#op-make-cred)

### LINKED ISSUE #559: Not clear what's executed in parallel in Section 4.1.3, Step 24.3
[Step 24.3](https://github.com/w3c/webauthn/blob/14c2733ca6a4a9568e4c48fef1b870448818e811/index.bs#L683) in Section 4.1.3 starts with "In parallel", but only mentions one operation. It's not clear what's supposed to be executed in parallel.

### PR DESCRIPTION
**Pull request by equalsJeffH**
this builds on PRs #371 and #495. it resolves these issues: 
improves #254 (#getAssertion section will be updated in a separate PR and issue #254 will be fixed at that time)
fix #466 
fix #472 
fix #536 
fix #559

***
[Preview](https://s3.amazonaws.com/pr-preview/w3c/webauthn/jeffh-fixup-algs-contd-3.html) | [Diff](https://s3.amazonaws.com/pr-preview/w3c/webauthn/b8c6027...4b2193e.html)

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -825,23 +831,45 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
+                        |distinctTransports| due to the properties of [=ordered sets=].
+
+                1. If |distinctTransports|
+                    <dl class="switch">
+                        :   [=list/is not empty=]
+                        ::  The client selects one |transport| value from |distinctTransports|, possibly incorporating local
+                            configuration knowledge of the appropriate transport to use with |authenticator| in making its
+                            selection.
+
+                            Then, using |transport|, invoke [=in parallel=] the [=authenticatorGetAssertion=] operation on
```
Comment (equalsJeffH) [inline index.bs]: I am thinking that the "in parallel" clauses in these substeps (of current step 16) ought to be replaced by a single "in parallel" clause on Step 16 itself ([new line 822](https://github.com/w3c/webauthn/pull/498/files#diff-ec9cfa5f3f35ec1f84feb2e59686c34dR822), above). 

If others agree, will do that in a subsequent commit.

File: index.bs [hunk 2]
```diff
@@ -855,28 +883,29 @@ When this method is invoked, the user agent MUST execute the following algorithm
 
     <dl class="switch">
```
Comment (equalsJeffH) [inline index.bs]: [new line 881](https://github.com/w3c/webauthn/pull/498/files#diff-ec9cfa5f3f35ec1f84feb2e59686c34dR881) above should be a substep of the step above it at [new line 878](https://github.com/w3c/webauthn/pull/498/files#diff-ec9cfa5f3f35ec1f84feb2e59686c34dR878). If others agree, will make that change in a subsequent commit.

File: index.bs (hunk 1 above)
Comment (equalsJeffH) [inline index.bs]: hold on there pardner, see @dominc's thoughts in this comment on issue #472: https://github.com/w3c/webauthn/issues/472#issuecomment-309871552

File: index.bs (hunk 2 above)
Comment (equalsJeffH) [inline index.bs]: I'm thinking this is a "go" given @dominc's thoughts in this comment on issue #472: https://github.com/w3c/webauthn/issues/472#issuecomment-309871552 and other examples of how algs are written, e.g. [in credman](https://w3c.github.io/webappsec-credential-management/#algorithm-request).

**Comment by equalsJeffH** [comment]
See @domenic's [comments in issue #472](https://github.com/w3c/webauthn/issues/472#issuecomment-309871552).  summary: this alg needs more work to get the thread-handling and JS-object construction stuff correct.

**Comment by AngeloKai** [comment]
Based on domenic's comment, there are still some works on how to address exception and creation of exception object. @equalsJeffH do you plan to do that in a separate PR? Or do you prefer to address them in this one altogether?

**Comment by equalsJeffH** [comment]
@AngeloKai asked
> Based on domenic's [comment](https://github.com/w3c/webauthn/issues/472#issuecomment-309871552), there are still some works on how to address exception and creation of exception object. @equalsJeffH do you plan to do that in a separate PR? Or do you prefer to address them in this one altogether?

thx, nominally I'd prefer to do it all in the context of this PR, unless there's some really good reasons to merge this as-is.

File: index.bs [hunk 3]
```diff
@@ -647,8 +653,8 @@ When this method is invoked, the user agent MUST execute the following algorithm
         1. If <code>|C|.{{transports}}</code> [=list/is not empty=], and |authenticator| is connected over a transport not
             mentioned in <code>|C|.{{transports}}</code>, the client MAY [=continue=].
         1. Otherwise, [=list/Append=] |C| to |excludeCredentialDescriptorList|.
-    1. [=In parallel=], invoke the [=authenticatorMakeCredential=] operation on |authenticator| with |rpId|,
-        |clientDataHash|, |options|.{{MakeCredentialOptions/rp}}, |options|.{{MakeCredentialOptions/user}},
+    1. [=In parallel=], invoke the [=authenticatorMakeCredential=] operation on |authenticator| with
+        |clientDataHash|, <code>|options|.{{MakeCredentialOptions/rp}}</code>, <code>|options|.{{MakeCredentialOptions/user}}</code>,
         |normalizedParameters|, |excludeCredentialDescriptorList|, and |authenticatorExtensions| as parameters.
     1. [=set/Append=] |authenticator| to |issuedRequests|.
```
Comment (rlin1) [inline index.bs]: I would typically do this step before invoking the authenticatorMakeCredential in parallel in order to avoid race conditions, in which the authenticatorMakeCredential already returns and hence would invoke the "authenticator indicates success" code without having this task in the list.

File: index.bs [hunk 4]
```diff
@@ -825,23 +831,45 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
             :   [=list/is empty=]
             ::  Using local configuration knowledge of the appropriate transport to use with |authenticator|, invoke
                 [=in parallel=] the [=authenticatorGetAssertion=] operation on |authenticator| with |rpId|, |clientDataHash|,
                 and |clientExtensions| as parameters.
+
+                Note: In this case, the [=[RP]=] did not supply a list of acceptable credential descriptors. Thus the
+                    authenticator is being asked to exercise any credential it may possess that is bound to the [=[RP]=].
         </dl>
 
     1. [=set/Append=] |authenticator| to |issuedRequests|.
```
Comment (rlin1) [inline index.bs]: see my previous command on reversing the order to appending to issuedRequests and invoking the request in parallel (and potential race conditions).

File: index.bs (hunk 1 above)
Comment (rlin1) [inline index.bs]: I think it is fine as it is right now.  Don't see the benefit of replacing the three in parallels with a single one.
Not that concrete implementations could still use a single in parallel invocation of authenticatorGet line-of-code and only prepare the parameters in the current is-empty/is-not-empty branches (as that is semantically equivalent).  Even code generators might automatically perform this transformation.

File: index.bs (hunk 3 above)
Comment (equalsJeffH) [inline index.bs]: hm, ok, what do others think, eg @jyasskin @jcjones ?

File: index.bs (hunk 1 above)
Comment (equalsJeffH) [inline index.bs]: actually, as suggested in https://github.com/w3c/webauthn/issues/472#issuecomment-309871552 (mentioned above), we will remove those [=in parallel=] invocations because we are already on a background thread given how we are invoked from https://w3c.github.io/webappsec-credential-management/#algorithm-request

**Comment by rlin1** [comment]
I think we found another limitation of the algorithm specified in section 4.1.4:
If the authenticator is used as first multifactor, i.e. user not yet known (e.g. no password auth, no cookies, ...), then the current algorithm sends no credentialIDs to the authenticator.
That is exactly the right thing to do.  If the authenticator has a UI, the Authenticator could ask the user to select one of the existing credentialIDs for the specific rpID.

However, if the authenticator hs no UI (e.g. like PQI My Lockey 360°), the authenticator would need platform support for the credentialID selection.

It would be great to add such support (e.g. by (1) letting the platform know whether or not the authenticator has a UI and (2) by letting the platform retrieve a list of credentialIDs from the authenticator - after user verification).  If we feel that this is too much of a heavy lifting at this stage, we should at least document the supported scenarios and the limitations.  At this stage it is not very obvious for a reader.

**Comment by equalsJeffH** [comment]
fix #520 in here also.

**Comment by equalsJeffH** [comment]
Ok, @mikewest has done an initial review of the related credman PR w3c/webappsec-credential-management#100 and seems to think it is nominally ok, so please also review the current state of this PR. Please note that it presently only addresses the #createCredential alg, #getAssertion is next on the list, and having early feedback will hopefully help the latter get done in a more polished fashion earlier.  thanks!

**Comment by equalsJeffH** [comment]
Ok, I have made further changes to accommodate @mikewest s initial feedback, which he gave over in the associated credman PR: https://github.com/w3c/webappsec-credential-management/pull/100#pullrequestreview-58690526

Depending on what we decide over there in credman wrt whether we add an `origin` member to each of [CredentialRequestOptions Dictionary](https://w3c.github.io/webappsec-credential-management/#credentialrequestoptions-dictionary) and [CredentialCreationOptions Dictionary](https://w3c.github.io/webappsec-credential-management/#credentialcreationoptions-dictionary), there may be further commits to add to this PR. C.f.: my comment: https://github.com/w3c/webappsec-credential-management/pull/100#issuecomment-327968338

With that in mind, review of the present state of this PR is welcome.  thanks :)

**Comment by nadalin** [comment]
@equalsJeffH any updates for the call tomorrow ?

**Comment by equalsJeffH** [comment]
@nadalin:
> any updates for the call tomorrow ?

credman [PR#100](https://github.com/w3c/webappsec-credential-management/pull/100) is interlinked with this webauthn PR #498

no response as yet to my request for review on PR#100 https://github.com/w3c/webappsec-credential-management/pull/100#issuecomment-327968338,  nor on comment above https://github.com/w3c/webauthn/pull/498#issuecomment-327969251

I also sent email: please review credman PR#100 and webauthn PR#498.. (fixup algs contd 3)
but no response as yet.  See my above-linked comments in credman PR#100 and webauthn PR#498 for details on what seems to still be open questions on the PRs.

**Comment by equalsJeffH** [comment]
I also need to grok and address @rlin1's comment above: https://github.com/w3c/webauthn/pull/498#issuecomment-319595322  -- will do that while awaiting review of present state by reviewers...

**Comment by rlin1** [comment]
The comment https://github.com/w3c/webauthn/pull/498#issuecomment-319595322 relates to issue #290 in FIDO CTAP (that one present an alternative approach).

**Comment by equalsJeffH** [comment]
@rlin wrote in https://github.com/w3c/webauthn/pull/498#issuecomment-330918874
> The comment [#issuecomment-319595322](https://github.com/w3c/webauthn/pull/498#issuecomment-319595322) relates to issue #290 in FIDO CTAP (that one present an alternative approach).

two things: 

1. #issuecomment-319595322](https://github.com/w3c/webauthn/pull/498#issuecomment-319595322) does not seem to relate to the thrust/theme of this PR #498. if i understand correctly, it is requesting a feature down at the authenticator API level.  Perhaps it should be submitted as a separate issue -- on CTAP? on the webauthn "aiuthenticator model" ? both?

2. I do not understand the "relates to issue #290 in FIDO CTAP (that one present an alternative approach)."  both webauthn issue #290 and [CTAP issue #290](https://github.com/fido-alliance/fido-2-specs/issues/290) seem unrelated.

File: index.bs [hunk 5]
```diff
@@ -737,6 +766,51 @@ authorizing an authenticator.
 </div>
 
 
+#### Construct the credential - `constructCredentialCallback` method #### {#constructCredentialCallback}
```
Comment (selfissued) [inline index.bs]: I forget the rationale for why this new method is being added.  Can you remind me?

**Comment by equalsJeffH** [comment]
I rebased this `jeffh-fixup-algs-contd-3` branch onto the current state of `master`.  apparently need to fix some issues for it to build correctly. apologies, will do so soon.

File: index.bs [hunk 6]
```diff
@@ -675,59 +687,76 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
-        |normalizedParameters|, |excludeCredentialDescriptorList|, and |authenticatorExtensions| as parameters.
+        |credTypesAndPubKeyAlgs|,
+        |excludeCredentialDescriptorList|,
+        and |authenticatorExtensions| as parameters.
+
     1. [=set/Append=] |authenticator| to |issuedRequests|.
 
-1. Start a timer for |adjustedTimeout| milliseconds. Then execute the following steps [=in parallel=]. The [=task source=] for
-    these [=tasks=] is the [=dom manipulation task source=].
+1. Start a timer for |adjustedTimeout| milliseconds.
```
[the remark this answers is not in the document]
Comment (equalsJeffH) [reply index.bs]: Yes, see all the issues linked in this page whose titles contain "parallel".  See also credman PR w3c/webappsec-credential-management#100. short answer: we are already "in parallel" when our #createCredential and #getAssertion algs are invoked from credman.  hence inappropriate to go "in parallel" in our algs.

File: index.bs (hunk 5 above)
Comment (equalsJeffH) [reply index.bs]: because in order to create an object associated with the current global object, we need to be on the "main thread" and so our algs, which have been invoked on a background thread by credman (aka "in parallel"), each return a callback (i've only done #createCredential so far) to credman's alg, who then [queues a task](https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-task) (ie on the main thread) to invoke the callback. see also: https://www.w3.org/2001/tag/doc/promises-guide/#queue-tasks

**Comment by equalsJeffH** [comment]
Bikeshed issues are resolved. 

Please note that this PR is **not yet ready-to-merge**.  I need to (at least) apply changes similar-to-those-in-#createCredential  to the #getAssertion alg. Also, this PR plus w3c/webappsec-credential-management#100 are inter-related and I have open questions to reviews such as @mikewest, @domenic, @battre.  However, if we do not receive review soon, we ought to just merge this once it is ready and go from there.

File: README.md [hunk 7]
```diff
@@ -45,16 +45,41 @@ This repository uses `.deploy-output.sh` to generate the Editor's Draft on the `
 # Creating a new Working Draft
 
 To build a new WD and upload it to the W3C publishing system:
-- Make sure Bikeshed is installed locally
+- Register as a W3C member and join the Web Authentication working group. Note down your W3C USERNAME and PASSWORD to use in the command below. If you don't remember either one, please go [here](https://www.w3.org/accounts/recover) to retrieve it.
```
Comment (jyasskin) [inline README.md]: I think this was accidentally pulled in during one of the historical rebases. Could you double-check all the changes on this branch and make sure they're related to the change you intend to make?

In particular, I think merge commits in the jeffh-fixup-algs-contd-3 branch are likely to confuse things.

File: index.bs [hunk 8]
```diff
@@ -1,9 +1,19 @@
-<h1>Web Authentication:<br>An API for accessing Public Key Credentials<br>[Level 1]</h1>
+<!--   ***********  Web Authentication - Level 1 - Spec source file ***********
```
Comment (jyasskin) [inline index.bs]: This change doesn't seem related to algorithm changes. Make it a separate PR.

File: index.bs [hunk 9]
```diff
@@ -579,38 +589,43 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
+            <code>|options|.{{MakePublicKeyCredentialOptions/rp}}.{{PublicKeyCredentialEntity/id}}</code> when calling
             {{CredentialsContainer/create()}}.
     </li>
 
-1. Let |normalizedParameters| be a new [=list=] whose [=list/items=] are pairs of {{PublicKeyCredentialType}} and
-    an {{AlgorithmIdentifier}}.
+1. Let |credTypesAndPubKeyAlgs| be a new [=list=] whose [=list/items=] are pairs of {{PublicKeyCredentialType}} and
+    a {{COSEAlgorithmIdentifier}}.
+
+1. [=list/For each=] |current| of <code>|options|.{{MakePublicKeyCredentialOptions/pubKeyCredParams}}</code>:
```
Comment (jyasskin) [inline index.bs]: I'd love if the additions of `<code>` were in a separate PR so I could focus on semantic changes in this one. It's not worth changing that now, but in the future please try.

File: index.bs [hunk 10]
```diff
@@ -671,63 +686,80 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
-            :   {{PublicKeyCredential/clientExtensionResults}}
-            ::  A new {{AuthenticationExtensions}} object containing the [=extension identifier=] → [=client extension output=]
-                entries created by running each extension's [=client extension processing=] algorithm to create the [=client
-                extension outputs=], for each [=client extension=] in {{AuthenticatorResponse/clientDataJSON}}.clientExtensions.
-
-        5. [=set/For each=] remaining |authenticator| in |issuedRequests| invoke the [=authenticatorCancel=] operation on
-            |authenticator| and [=set/remove=] it from |issuedRequests|.
-        6. Return |value| and terminate this algorithm.
-    </dd>
+        :   If the |adjustedTimeout| timer expires,
```
Comment (jyasskin) [inline index.bs]: I'm not manually-diffing this section, though. Please pull the editorial `<dt>`->`:` change into a separate PR so it's clear what's actually changed here.

File: index.bs [hunk 11]
```diff
@@ -737,6 +769,51 @@ authorizing an authenticator.
...
+
+This algorithm accepts one argument:
+
+    :   |global|
+    ::  Which must be its caller's [=current settings object=]'s [=environment settings object/global object=].
+
+When this method is invoked, the user agent MUST execute the following algorithm:
+
+1.  Let |attestationObject| be a new {{ArrayBuffer}}, created using |global|'s [=%ArrayBuffer%=], containing the
+    bytes of |credentialCreationData|.<code>[=attestationObjectResult=]</code>'s value.
```
Comment (jyasskin) [inline index.bs]: This `<code>` should be before `|credentialCreationData|`, right?

File: index.bs [hunk 12]
```diff
@@ -737,6 +769,51 @@ authorizing an authenticator.
...
+
+3.  Let |id| be |attestationObject|.<code>authData.attestation data.credential ID</code> (see [[#sec-attestation-data]]
```
Comment (jyasskin) [inline index.bs]: Fun that our field names have spaces in them.

File: index.bs [hunk 13]
```diff
@@ -737,6 +769,51 @@ authorizing an authenticator.
...
+
+        :   {{AuthenticatorResponse/clientDataJSON}}
+        ::  A new {{ArrayBuffer}}, created using |global|'s [=%ArrayBuffer%=], containing the bytes of
+            |credentialCreationData|.<code>[=clientDataJSONResult=]</code>.
+
+        :   {{AuthenticatorAttestationResponse/attestationObject}}
+        ::  |attestationObject|
+
+    :   {{PublicKeyCredential/clientExtensionResults}}
+    ::  A new {{AuthenticationExtensions}} object associated with |global| containing the bytes
```
Comment (jyasskin) [inline index.bs]: The record would just contain the value of extensionOutputsMap, not its bytes.

File: index.bs (hunk 8 above)
Comment (equalsJeffH) [reply index.bs]: well, we have done such while-i'm-in-there editors' notes since we began this project and if we waited to do separate PRs for ever little thing they'd never get done.  if it's innocuous and not obviously incorrect, then we've been incorporating them.

File: index.bs (hunk 9 above)
Comment (equalsJeffH) [reply index.bs]: please see comment above.

File: index.bs (hunk 10 above)
Comment (equalsJeffH) [reply index.bs]: well, I have manually diffed it for you as a [rendered HTML diff here](http://kingsmountain.com/doc/diff/diff-webauthn-index-jeffh-fixup-algs-contd-3-a1b998f--from--master-3ee8ed5.html#createCredential), and I will do a text-only kdiff3 one tomorrow morning (i at least find the latter useful).   given above comment wrt while-we're-in-there markup improvements, I am loath (and do not have the time) to put all the <dd>...</dd> crud back in there.

File: index.bs (hunk 12 above)
Comment (equalsJeffH) [reply index.bs]: yeah, that's because they are informal on-the-fly names, which need to be formalized, pls see issue #233 subissues (2) & (3)   ("formalized" whether by adding a "name" column to the tables, or by using ABNF, or whatever)

File: README.md (hunk 7 above)
Comment (equalsJeffH) [reply README.md]: > double-check all the changes on this branch and make sure they're related to the change you intend to make?

done via the diff I link-to above.  looks fine to me.  I had to rebase onto master because master has been changing out from under this branch. 

> In particular, I think merge commits in the jeffh-fixup-algs-contd-3 branch are likely to confuse things.

am not sure what you mean by this.

File: index.bs (hunk 9 above)
Comment (jyasskin) [reply index.bs]: For these in particular, I'd be totally happy if you just merge them without review. I just don't want to see them when I'm trying to review something else.

File: index.bs (hunk 10 above)
Comment (equalsJeffH) [reply index.bs]: here's a [rendered side-by-side textual diff](http://kingsmountain.com/doc/diff/diff-webauthn-index-jeffh-fixup-algs-contd-3-a1b998f--from--master-3ee8ed5.text.pdf) (between this branch and the current state of the master branch) with which, in combination with the [rendered HTML daisydiff](http://kingsmountain.com/doc/diff/diff-webauthn-index-jeffh-fixup-algs-contd-3-a1b998f--from--master-3ee8ed5.html#createCredential), one can clearly see what has changed. daisydiff seems to get confused by things such as the "switch" stuff and so the textual diff helps to see what a reader will see.

File: index.bs (hunk 9 above)
Comment (equalsJeffH) [reply index.bs]: ok, well, sorry. i guess trying to rely on reviewing plain source html complicates things. and the [built-in diff up at the top in the orig post](https://s3.amazonaws.com/pr-preview/w3c/webauthn/cca20d3...a1b998f.html) does not diff against the current state of master (I just checked (it would be more helpful if it would)). That's why I generate rendered html diffs as needed (links below).

File: index.bs (hunk 11 above)
Comment (equalsJeffH) [reply index.bs]: yes, thx

File: index.bs (hunk 9 above)
Comment (equalsJeffH) [reply index.bs]: Oh, also, in looking at the built-in github;com plain html side-by-side diff, it seems it is diffing against an older commit of master and so there's stuff in here that's not relevant (sorry), so using the rendered html diffs I generated (links below) are the way to go.

File: index.bs (hunk 13 above)
Comment (equalsJeffH) [reply index.bs]: ok, thx, so it seems i can just delete "the bytes of" ?

**Comment by equalsJeffH** [comment]
Unfortunately, rebasing on the current state of master messed up the auto-generated github diffs because it seems they are diff'g against an older commit rather than the one that is the current state of master. 

So, here's a [rendered side-by-side textual diff](http://kingsmountain.com/doc/diff/diff-webauthn-index-jeffh-fixup-algs-contd-3-a1b998f--from--master-3ee8ed5.text.pdf) (between this branch and the current state of the master branch) with which, in combination with the [rendered HTML daisydiff](http://kingsmountain.com/doc/diff/diff-webauthn-index-jeffh-fixup-algs-contd-3-a1b998f--from--master-3ee8ed5.html), one can clearly see what has changed (so far) wrt the current state of master. I will endeavor to produce such diffs as we work this PR to completion.

File: index.bs (hunk 3 above)
Comment (equalsJeffH) [reply index.bs]: plus, this entire alg is now running "in parallel" and the `[=in parallel=]` on old line 678 is now gone.  So I'm not sure there's anything to change here.

File: index.bs (hunk 13 above)
Comment (equalsJeffH) [reply index.bs]: i deleted "the bytes of".

**Comment by equalsJeffH** [comment]
new, improved "return an algorithm" technique is ready to review. 
Rendered HTML and text-only diffs from current state of master branch are here:

rendered html: http://kingsmountain.com/doc/diff/diff-webauthn-index-jeffh-fixup-algs-contd-3-7b272f1--from--index-master-ee174c2.html

rendered text-only: http://kingsmountain.com/doc/diff/diff-webauthn-index-jeffh-fixup-algs-contd-3-7b272f1--from--index-master-ee174c2.html.pdf

File: index.bs [hunk 14]
```diff
@@ -701,62 +707,103 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
+                ::  whose value is the bytes of |clientDataJSON|.
+
+                :   <code><dfn>extensionOutputsMap</dfn></code>
+                ::  whose value is an [=ordered map=] with [=map/keys=] of type [=extension identifier=]
+                    and [=map/values=] of type [=client extension output=]. <code>[=extensionOutputsMap=]</code>'s
+                    [=map/entries=] are created by running each extension's [=client extension processing=] algorithm to create
+                    the [=client extension outputs=], for each [=client extension=] in
+                    <code>{{AuthenticatorResponse/clientDataJSON}}.clientExtensions</code>.
+
+            1.  Let |constructCredentialAlg| be an algorithm that takes a [=environment settings object/global object=]
```
Comment (jyasskin) [inline index.bs]: `takes a [=global object=]`, which should resolve to https://html.spec.whatwg.org/multipage/webappapis.html#global-object.

File: index.bs [hunk 15]
```diff
@@ -701,62 +707,103 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
+            2. [=set/For each=] remaining |authenticator| in |issuedRequests| invoke the [=authenticatorCancel=] operation on
+                |authenticator| and [=set/remove=] it from |issuedRequests|.
+
+        :   If any |authenticator| returns an error status,
+        ::  [=set/Remove=] |authenticator| from |issuedRequests|.
+
+        :   If any |authenticator| indicates success,
+        ::  1.  [=set/Remove=] |authenticator| from |issuedRequests|.
+
+            1.  Let |credentialCreationData| be a [=struct=] whose [=items=] are:
```
Comment (jyasskin) [inline index.bs]: This is fine, but I believe you'd also be fine just defining three variables with the values you've put into items here, and doing that would be a bit shorter.

File: index.bs [hunk 16]
```diff
@@ -701,62 +707,103 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
+                            {{AuthenticationExtensions}}'s type.
+
+                1.  Return |pubKeyCred|.
+
+            1. [=set/For each=] remaining |authenticator| in |issuedRequests| invoke the [=authenticatorCancel=] operation on
+                |authenticator| and [=set/remove=] it from |issuedRequests|.
+
+            1. Return |constructCredentialAlg| and terminate this algorithm.
+
+            Issue: Do we need to replicitly return both |constructCredentialAlg| and |credentialCreationData| here?
```
Comment (jyasskin) [inline index.bs]: I believe that it's sufficient to return only *constructCredentialAlg*, on the theory that algorithm variables are captured by any subroutines that use them and only garbage-collected when all capturing routines are done. The main thing you have to guarantee is that no variable is used and modified in parallel, which we have to do by inspection whether or not you explicitly return the set of captured variables.

File: index.bs (hunk 15 above)
Comment (equalsJeffH) [reply index.bs]: Ok, tho will leave this as-is for now since it is OK and we want to get this merged.

File: index.bs [hunk 17]
```diff
@@ -701,62 +707,103 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
+                ::  whose value is the bytes returned from the successful [=authenticatorMakeCredential=] operation.
+
+                    Note: this value is <code>attObj</code>, as defined in [[#generating-an-attestation-object]].
+
+                :   <code><dfn>clientDataJSONResult</dfn></code>
+                ::  whose value is the bytes of |clientDataJSON|.
+
+                :   <code><dfn>extensionOutputsMap</dfn></code>
+                ::  whose value is an [=ordered map=] with [=map/keys=] of type [=extension identifier=]
+                    and [=map/values=] of type [=client extension output=]. <code>[=extensionOutputsMap=]</code>'s
```
[the remark this answers is not in the document]
Comment (equalsJeffH) [reply index.bs]: good catch, IIUC.  Tho, this parag will now different because we merged PR #633 -- am updating this parag accordingly

[the remark this answers is not in the document]
Comment (equalsJeffH) [reply index.bs]: well, actually ended up using the "old wording" you referred to.  Tho this raises issue #657.

File: index.bs (hunk 14 above)
Comment (equalsJeffH) [reply index.bs]: ok.

File: index.bs (hunk 16 above)
Comment (equalsJeffH) [reply index.bs]: ok, thanks, I'll remove that "Issue:"

**Comment by equalsJeffH** [comment]
per 18-Oct webauthn call, I've addressed @jyasskin's latest comments.  This is ready to merge AFAIK.  Since I did make some modest changes in response to the latter comments, perhaps someone(s) oughta review prior to merge? 

Making similar changes to the #getAssertion section is yet to do in a separate soon-to-follow PR. 

Regarding [Rolf's oldish review](https://github.com/w3c/webauthn/pull/498#pullrequestreview-46490753) that is requesting changes, It seems to me that [there is nothing to do in response to it, unless I am misunderstanding things](https://github.com/w3c/webauthn/pull/498#issuecomment-332398969). [I ping'd him on-list](https://lists.w3.org/Archives/Public/public-webauthn/2017Sep/0458.html) a few weeks ago about this, but have not heard back.  So are we ok to "dismiss" is review in order to merge this?

thanks!
